Expose which rule labels apply to an endpoint

raybejjani opened the pull request:
**Summary of changes**:
The labels of rules that generated an endpoint policy are now included in the endpoint object returned by the API (in `endpoint get` and `endpoint list`). The labels are stored separately for ingress and egress for each of L3 (CIDRPolicy) and L4/L7. 
A follow up PR ( https://github.com/cilium/cilium/pull/2479) will add some output to cilium endpoint list (alas, our tests are brittle and that PR is large enough to be separate). It will also include a docs update to use the new fields.

Fixes: #2184 

```release-note
cilium endpoint get reports which policy rule labels are in effect for an endpoint
```

**How to test (optional)**:
- setup some endpoints, add some policies
```
kubectl create -f examples/minikube/demo.yaml
kubectl create -f examples/minikube/l3_l4_policy.yaml -f examples/minikube/l3_l4_l7_policy.yaml
```

- `cilium endpoint get`

```
cilium endpoint get 51796 -o json
...
     "cidr-policy": {
        "egress": [
          {
            "derived-from-rules": [
              [
                "unspec:io.cilium.k8s-policy-name=cidr-rule",
                "unspec:io.cilium.k8s-policy-namespace=default"
              ]
            ],
            "rule": "20.1.1.1/32"
          },
          {
            "derived-from-rules": [
              [
                "unspec:io.cilium.k8s-policy-name=cidr-rule",
                "unspec:io.cilium.k8s-policy-namespace=default"
              ]
            ],
            "rule": "10.128.0.0/9"
          },
          {
            "derived-from-rules": [
              [
                "unspec:io.cilium.k8s-policy-name=cidr-rule",
                "unspec:io.cilium.k8s-policy-namespace=default"
              ]
            ],
            "rule": "10.0.0.0/10"
          },
          {
            "derived-from-rules": [
              [
                "unspec:io.cilium.k8s-policy-name=cidr-rule",
                "unspec:io.cilium.k8s-policy-namespace=default"
              ]
            ],
            "rule": "10.64.0.0/11"
          },
          {
            "derived-from-rules": [
              [
                "unspec:io.cilium.k8s-policy-name=cidr-rule",
                "unspec:io.cilium.k8s-policy-namespace=default"
              ]
            ],
            "rule": "10.112.0.0/12"
          }
        ],
        "ingress": []
      },
...
      "l4": {
        "egress": [],
        "ingress": [
          {
            "derived-from-rules": [
              [
                "unspec:io.cilium.k8s-policy-name=access-backend",
                "unspec:io.cilium.k8s-policy-namespace=default"
              ]
            ],
            "rule": "{\n  \"port\": 80,\n  \"protocol\": \"TCP\"\n}"
          }
        ]
      }
    },
...
```

raybejjani committed:
api: Include L3 rule labels in endpoint info

The labels of L3 rules that caused the L3 rule to be applied  are
included in the API endpoint object, attached to the ingress/egress rule
they impacted.

Signed-off-by: Ray Bejjani <ray@covalent.io>

raybejjani committed:
api: Include L4 rule labels in endpoint info

The labels of  L4/L7 rules that caused that rule to be applied are
included in the API endpoint object, attached to the ingress/egress rule
they impacted.

Signed-off-by: Ray Bejjani <ray@covalent.io>

Switch to L4+rules attribution

policy: Undo Refactor L4PolicyMap to be a struct

manalibhutiyani commented on pkg/policy/l4.go:
```diff
@@ -254,7 +258,7 @@ func (l4 L4PolicyMap) HasRedirect() bool {
 // Otherwise, returns api.Allowed.
 func (l4 L4PolicyMap) containsAllL3L4(labels labels.LabelArray, ports []*models.Port) api.Decision {
-	if len(l4) == 0 {
+	if len(l4.Filters) == 0 {
```
Do we need to also check for `if len(l4) == 0 {`  before checking for l4.filters and return some error if its nil?

aanm replied:
@manalibhutiyani `l4` is no longer a map.

aanm commented on pkg/policy/l4.go:
```diff
@@ -230,12 +230,16 @@ func (l4 L4Filter) matchesLabels(labels labels.LabelArray) bool {
 // L4PolicyMap is a list of L4 filters indexable by protocol/port
 // key format: "port/proto"
-type L4PolicyMap map[string]L4Filter
+type L4PolicyMap struct {
```
This is no longer a map, the structure documentation is not correct. I would suggest a refactor of the struct's name

raybejjani replied:
oh, I didn't think we called it Map because it was a go map (L3Policy map is also not a map). I actually thought it described the usage as a mapping of proto/port to a rule. I would prefer to drop the Map in both L4PolicyMap and L3PolicyMap if they don't mean anything.

aanm commented on pkg/labels/labels.go:
```diff
@@ -164,6 +164,14 @@ func (l Labels) String() string {
+// String representation
+func (l Labels) AddFromLabelArray(lbls LabelArray) {
+	for _, lbl := range lbls {
+		l[lbl.String()] = lbl
```
This is not respecting the documentation of the Labels struct:

https://github.com/cilium/cilium/blob/5fc5b37a2b9d31da0bef6ea61eaac63f7fdf24ff/pkg/labels/labels.go#L151-L152


The map's key should be the same as the labels' key.

aanm commented on pkg/k8s/network_policy_v1beta_test.go:
```diff
@@ -107,15 +107,24 @@ func (s *K8sSuite) TestParseNetworkPolicyDeprecated(c *C) {
+				},
+			},
+			SourceRuleLabels: labels.Labels{
+				"unspec:io.cilium.k8s-policy-name":              labels.ParseLabel("unspec:io.cilium.k8s-policy-name"),
```
This is not respecting the documentation of the Labels struct:

https://github.com/cilium/cilium/blob/5fc5b37a2b9d31da0bef6ea61eaac63f7fdf24ff/pkg/labels/labels.go#L151-L152


The map's key should be the same as the labels' key.

aanm commented on pkg/k8s/network_policy_test.go:
```diff
@@ -119,15 +119,24 @@ func (s *K8sSuite) TestParseNetworkPolicy(c *C) {
+				},
+			},
+			SourceRuleLabels: labels.Labels{
+				"unspec:io.cilium.k8s-policy-namespace=default": labels.ParseLabel("unspec:io.cilium.k8s-policy-namespace=default"),
```
This is not respecting the documentation of the Labels struct:

https://github.com/cilium/cilium/blob/5fc5b37a2b9d31da0bef6ea61eaac63f7fdf24ff/pkg/labels/labels.go#L151-L152


The map's key should be the same as the labels' key.

aanm commented on pkg/policy/l3.go:
```diff
@@ -32,9 +33,11 @@ import (
+	IPv6Count        int                  // Count of IPv6 prefixes in 'Map'
+	IPv4Count        int                  // Count of IPv4 prefixes in 'Map'
+	SourceRuleLabels labels.Labels        // The User rule labels that resulted in this Map. Can be empty
+	SourceRuleCount  int64                // Number of rules that resulted in this Map
```
Out of curiosity, why is the count important?

raybejjani replied:
I added this once I realised that many rules will have no labels, so it was impossible to get a quick count of rules affecting an endpoint. This might be a bad choice, though, if we prefer to have the user only ever use `cilium policy` to get policies or their counts. From some quick testing with `cilium policy get` if you pass no labels you get all rules, not ones without labels :/

aanm commented on pkg/policy/l4.go:
```diff
@@ -230,12 +230,16 @@ func (l4 L4Filter) matchesLabels(labels labels.LabelArray) bool {
-type L4PolicyMap map[string]L4Filter
+type L4PolicyMap struct {
+	Filters          map[string]L4Filter
+	SourceRuleLabels labels.Labels // The User rule labels that resulted in this Map. Can be empty
```
I've noticed you've used `SourceRuleLabels` and `SourceRuleCount` on a previous structure (`L3PolicyMap`). Does it make sense to create a new structure and embedded that new structure in `L4PolicyMap` and `L3PolicyMap`? (This is just a thought, I'm ok as it is.)

raybejjani replied:
I think I will, since I can then attach methods to merge them and manage them. This is better since I'm violating the usage of the Labels type and I need to fix that too.

aanm commented on pkg/policy/l4.go:
```diff
@@ -230,12 +230,16 @@ func (l4 L4Filter) matchesLabels(labels labels.LabelArray) bool {
 
 // HasRedirect returns true if at least one L4 filter contains a port
 // redirection
 func (l4 L4PolicyMap) HasRedirect() bool {
```
Since `L4PolicyMap` is no longer a map, should we change the receiver to a pointer? Answer: Not on this case. But are there any other cases where there are methods that use attributes of `L4PolicyMap` and need a receiver as a pointer?

raybejjani replied:
You are correct overall, though. I'll change them all out to be consistent (it also means no one can accidentally call them on a non-pointer)

aanm commented on pkg/policy/rule.go:
```diff
@@ -263,25 +264,35 @@ func (r *rule) resolveL4Policy(ctx *SearchContext, state *traceState, result *L4
 		}
-		for _, r := range r.Ingress {
-			cnt, err := mergeL4(ctx, "Ingress", r.FromEndpoints, r.ToPorts, result.Ingress)
+		result.Ingress.SourceRuleLabels = make(labels.Labels)
```
I've a feeling `result` is being used multiple times by the caller of `resolveL4Policy`, aren't you going to overwrite the `SourceRuleLabels` on every loop?

https://github.com/cilium/cilium/blob/5fc5b37a2b9d31da0bef6ea61eaac63f7fdf24ff/pkg/policy/repository.go#L144-L154

raybejjani replied:
You are correct. In fact, the looping might include all endpoints so this is the wrong place to accumulate this information :(

raybejjani committed:
daemon: Add LabelArrayList, LabelArray.DeepCopy & .GetModel & nil safety

LabelArrayList is a slice of LabelArray objects. It's a little unwieldy
but is used to collect rule labels to send out via the API later.
The LabelArray.DeepCopy and .GetModel functions match equivalent ones on
labels.Labels, the other Label collection class.
All these DeepCopy functions can be called on a nil, avoiding nil checks
on all invocations of DeepCopy.

tgraf commented on api/v1/openapi.yaml:
```diff
@@ -974,11 +974,25 @@ definitions:
         items:
           type: string
+      ingress-source-rule-labels:
+        description: Labels of ingress L3 rules in effect on this endpoint
```
Is it feasible to do something like this instead?

    "l4": {
        "egress": [{
             "rule": "{\n  \"port\": 80,\n  \"protocol ...",
             "derived-from-rules": [{["foo", "bar"]},{["blah"]}]
        },{
            "rule": "{\n  \"port\": 80,\n  \"protocol ...",
            "derived-from-rules": [{["baz2"]}]
        }]
    }

aanm replied:
@raybejjani also, what's the output for `policy.allowed-consumers`?

raybejjani replied:
I don't think I changed allowed-consumers. A random endpoint outputs:
```json
    "policy": {
      "allowed-consumers": [
        1
      ],
```

raybejjani replied:
I'll merge the derived-from thing with rule. Are we ok changing the API this way? Do I need to roll a version somewhere?

tgraf replied:
> I'll merge the derived-from thing with rule. Are we ok changing the API this way? Do I need to roll a version somewhere?

No reason to bump pre 1.0

tgraf commented on api/v1/openapi.yaml:
```diff
@@ -960,11 +960,25 @@ definitions:
         type: array
         items:
           type: string
+      ingress-source-rule-labels:
```
Same here, I would rename this.

tgraf reviewed (changes requested):
Overall this looks great. I wonder whether we can provide more context here, see the comment inline

manalibhutiyani commented on pkg/policy/l4.go:
```diff
@@ -337,17 +346,27 @@ func (l4 *L4Policy) GetModel() *models.L4Policy {
-		Ingress: ingress,
-		Egress:  egress,
+		Ingress:                 ingress,
+		IngressSourceRuleLabels: ingressLabels,
```
`IngressSourceRuleLabels` and `EgressSourceRuleLabels ` . I am not sure we need `Source`   in the variable name. Seems confusing?

raybejjani replied:
I'm afraid these names have to match the name in the API json, mangled by whatever swagger does (which seems to snake-case it and remove the dashes). I will be changing them to be derived-from-rules in the api, so they would look like `DerivedFromRules` which is a little closer to what you suggested?

manalibhutiyani reviewed (approved):
Overall changes look good. Thanks for doing this! I do agree with Thomas's review suggestions about ```"rule": "{\n  \"port\": 80,\n  \"protocol ...",
         "derived-from-rules": [{["foo", "bar"]},{["blah"]}]
    },{```

tgraf commented on pkg/labels/array.go:
```diff
@@ -129,3 +129,30 @@ func (ls LabelArray) Get(key string) string {
+func (ls LabelArray) DeepCopy() LabelArray {
+	o := make(LabelArray, 0, len(ls))
+	for _, v := range ls {
+		if v == nil {
```
Should we make `func (l Labels) DeepCopy() Labels {` `nil` aware instead?

raybejjani replied:
It already is? in the sense that it won't crash with a nil. Did you mean DeepCopy on a nil would return a nil?
If you're wondering why I didn't use the Labels type since it has a DeepCopy already, it's because there is no function to turn a LabelArray into a Labels but the Rule objects have LabelArrays (so, I figured writing the copy would be closer to the what we had before than making a converter).

tgraf replied:
> It already is? in the sense that it won't crash with a nil. Did you mean DeepCopy on a nil would return a nil?

Yes, if you return nil on nil you don't need to embed the nil conditional everywhere

raybejjani replied:
done, I think

tgraf commented on pkg/policy/l3.go:
```diff
@@ -22,24 +22,32 @@ import (
+// L3PolicyMapRule holds a L3 prefix and the rule labels that allow it.
+type L3PolicyMapRule struct {
+	Prefix           net.IPNet
+	DerivedFromRules []labels.LabelArray
```
Let's declare a type for this. `LabelArrayList` or `LabeLArraySlice` or something else

raybejjani replied:
done

tgraf commented on pkg/policy/l3.go:
```diff
@@ -135,14 +151,32 @@ func (l3 *L3Policy) GetModel() *models.CIDRPolicy {
+		ruleLabels := make([][]string, 0, len(v.DerivedFromRules))
+		for _, lblArr := range v.DerivedFromRules {
+			ruleLabels = append(ruleLabels, lblArr.GetModel())
+		}
```
So we can move this code into the new type and share it.

tgraf commented on pkg/policy/l4.go:
```diff
@@ -336,14 +343,32 @@ func (l4 *L4Policy) GetModel() *models.L4Policy {
+		ruleLabels := make([][]string, 0, len(v.DerivedFromRules))
+		for _, lblArr := range v.DerivedFromRules {
+			ruleLabels = append(ruleLabels, lblArr.GetModel())
+		}
```
same here

tgraf commented on pkg/policy/l4.go:
```diff
@@ -336,14 +343,32 @@ func (l4 *L4Policy) GetModel() *models.L4Policy {
+		ruleLabels := make([][]string, 0, len(v.DerivedFromRules))
+		for _, lblArr := range v.DerivedFromRules {
+			ruleLabels = append(ruleLabels, lblArr.GetModel())
+		}
```
same here

tgraf reviewed (changes requested):
Output and structure looks great now! Small request to hide the slice of LabelArray behind a type. The naming gets a bit awkward at this point but I think it's better than open coding this everywhere.

ianvernon commented on api/v1/server/embedded_spec.go:
```diff
@@ -1709,6 +1709,25 @@ func init() {
+          }
+        },
+        "rule": {
+          "description": "The policy rule as json",
```
This description is self-referential; can you state what the rule will appear as, i.e. its name, etc.

raybejjani replied:
I totally missed this but I can fix it up in another PR. It isn't terribly useful as a description, but I'm not sure I understand what you mean by "what the rule will appear as". This type replaces what was a string array with the description "List of L4 ingress rules" (or L3). Did you something like this in mind?

The policy rule's metadata and L3/L4/L7 restrictions as derived from the union of the set of rules in the cilium agent

(I was trying to avoid using the word rule again, although this derived rule might not be 1:1 equivalent to something submitted by the user).

ianvernon commented on pkg/labels/array.go:
```diff
@@ -129,3 +129,30 @@ func (ls LabelArray) Get(key string) string {
+	return o
+}
+
+// GetModel returns the LabelArray as a string array with fully qualified labels.
```
Nit: fully-qualified

raybejjani replied:
done

ianvernon reviewed (approved):
Looks good overall. Regarding how this impacts #1781 - in that PR, I currently collapse the CIDRs used in the L3PolicyMap per-endpoint. Would you recommend that I create a duplicate L3PolicyMap that contains the coalesced CIDRs per endpoint so we can maintain this functionality? Or, I could add logic in the coalescing that tracks which rules correspond to which CIDR.
